### PR TITLE
Fix TransactionTooLargeException when opening HistoryDetailActivity

### LINKED ISSUE #16012: RuntimeException: Failure from system
**Sentry Url**: https://sentry.io/share/issue/7a2d9cc64ef945169866fd2d7fe42af6/
**User Count**: 323
**Count**: 886
**First Release**: 18.8-rc-2
**First Seen**: 2021-12-08T09:37:41.206000Z
**Last Seen**: 2022-02-28T06:39:16Z
**24 Hours**: 8
**30 Days**: 440

```
RuntimeException: Failure from system
     at org.wordpress.android.ui.ActivityLauncher.viewHistoryDetailForResult(ActivityLauncher.java:1102)
     at org.wordpress.android.ui.posts.EditPostActivity.onHistoryItemClicked(EditPostActivity.java:1937)
     at org.wordpress.android.ui.posts.HistoryListFragment.setObservers$lambda-7(HistoryListFragment.kt:155)
     at org.wordpress.android.ui.posts.HistoryListFragment.lambda$lEsuMbG6B9pBZTJlu9vkgLw-Wk4(null:0)
     at org.wordpress.android.ui.posts.-$$Lambda$HistoryListFragment$lEsuMbG6B9pBZTJlu9vkgLw-Wk4.onChanged(null:4)
     at org.wordpress.android.viewmodel.SingleLiveEvent.observe$lambda-0(SingleLiveEvent.kt:50)
     at org.wordpress.android.viewmodel.SingleLiveEvent.lambda$QDPJ8FUoo3cDlf_Lb7Y_qP14f6o(null:0)
     at org.wordpress.android.viewmodel.-$$Lambda$SingleLiveEvent$QDPJ8FUoo3cDlf_Lb7Y_qP14f6o.onChanged(null:4)
     at org.wordpress.android.viewmodel.SingleLiveEvent.setValue(SingleLiveEvent.kt:58)
     at org.wordpress.android.viewmodel.history.HistoryViewModel.onItemClicked(HistoryViewModel.kt:218)
...
(32 additional frame(s) were not displayed)
```

### PR DESCRIPTION
**Pull request by RenanLukas**
Fixes #16012

**Depends on: https://github.com/wordpress-mobile/WordPress-FluxC-Android/pull/2310**

To test:
1 - Log in
2 - Select a site
3 - On "My Site", tap on "Posts"
4 - Tap on a post that was updated at least once
5 - Tap on overflow menu (three vertical dots on top right)
6 - Tap on "History"
7 - Tap on any Revision on list. Revision details screen should open, instead of the app crashing.

## Regression Notes
1. Potential unintended areas of impact
Revision details ViewPager (introduced a bug while developing and then fixed it)

2. What I did to test those areas of impact (or what existing automated tests I relied on)
Manual testing opening history details screen with different sites, moving the ViewPager from beginning to end.

3. What automated tests I added (or what prevented me from doing so)
I've tried to update `HistoryViewModelTest` with my changes, but couldn't do it because `OnRevisionsFetched` visibility is limited. If there are any ideas to have this code tested the way it is, please let me know.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by RenanLukas** [comment]
👀 CI

File: WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java [hunk 1]
```diff
@@ -1941,8 +1941,20 @@ public void onBackPressed() {
     public void onHistoryItemClicked(@NonNull Revision revision, @NonNull List<Revision> revisions) {
         AnalyticsTracker.track(Stat.REVISIONS_DETAIL_VIEWED_FROM_LIST);
         mRevision = revision;
+        ActivityLauncher.viewHistoryDetailForResult(this, mRevision, getRevisionsIds(revisions));
+    }
 
-        ActivityLauncher.viewHistoryDetailForResult(this, mRevision, revisions);
+    private long[] getRevisionsIds(@NonNull final List<Revision> revisions) {
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostActivity.java:1950]: If I understand this correctly, number of returned ID's will be same as the number of revisions, so maybe we don't need to extract them into `idsList` first?

Can we simplify it into something like this (not tested)?:

```
  private long[] getRevisionsIds(@NonNull final List<Revision> revisions) {
        final long[] idsArray = new long[revisions.size()];
        for (int i = 0; i < revisions.size(); i++) {
            final Revision current = revisions.get(i);
            idsArray[i] = current.getRevisionId();
        }
        return idsArray;
    }
```

File: WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java [hunk 2]
```diff
@@ -172,6 +179,37 @@ public void onClick(View view) {
...
+
+    @Nullable
+    private ArrayList<Revision> mapRevisionModelsToRevisions(@Nullable final List<RevisionModel> revisionModels) {
+        if (revisionModels == null) {
+            return null;
+        }
+        final ArrayList<Revision> revisions = new ArrayList<>();
+        for (int i = 0; i < revisionModels.size(); i++) {
+            final RevisionModel current = revisionModels.get(i);
+            final Revision revision = new Revision(current.getRevisionId(), current.getDiffFromVersion(),
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java]: `Revision` has constructor that takes a `RevisionModel`. Wonder if we can utilize it?

File: WordPress/src/main/java/org/wordpress/android/viewmodel/history/HistoryViewModel.kt [hunk 3]
```diff
@@ -223,6 +223,18 @@ class HistoryViewModel @Inject constructor(
...
 
+    private fun saveRevisionsToLocalDB(post: PostModel, revisions: List<RevisionModel>) {
+        revisions.forEach {
+            postStore.setLocalRevision(it, site, post)
+        }
+    }
+
+    private fun removeRevisionsFromLocalDB(post: PostModel, revisions: List<RevisionModel>) {
+        revisions.forEach {
+            postStore.deleteLocalRevision(it, site, post)
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/viewmodel/history/HistoryViewModel.kt]: If we only remove the revisions received from the endpoint, there is a chance we might have some revisions not removed, if for some reason they are not in the response. What do you think about removing all the revision for specific post?

File: WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java [hunk 4]
```diff
@@ -172,6 +179,37 @@ public void onClick(View view) {
...
     }
 
+    private void mapRevisions() {
+        if (getArguments() != null) {
+            mRevision = getArguments().getParcelable(EXTRA_CURRENT_REVISION);
+
+            final long[] previousRevisionsIds = getArguments().getLongArray(EXTRA_PREVIOUS_REVISIONS_IDS);
+            final List<RevisionModel> revisionModels = new ArrayList<>();
+            for (final long revisionId : previousRevisionsIds) {
+                revisionModels.add(mPostStore.getRevisionById(revisionId));
```
Comment (khaykov) [inline WordPress/src/main/java/org/wordpress/android/ui/history/HistoryDetailContainerFragment.java]: This is probably not a significant issue, but I wonder if performance might suffer when accessing DB from main thread for a post wit a lot of changes.

**Review by khaykov — COMMENTED** [review]
Thanks for tackling this one, @renanferrari 🙇  It works well! I left couple of comments in the code.

**Comment by RenanLukas** [comment]
Thanks for the review @khaykov ! I've addressed the issues we discussed 👍  (except [this one](https://github.com/wordpress-mobile/WordPress-Android/pull/16105#discussion_r827331041) - if you think it's worth, I'll create a separate issue for it and will work on it in the future)

**Review by khaykov — APPROVED** [review]
Perfect, thank you! lgtm 🚀

**Pull request merged**